reconcilers/childset.go

```diff
@@ -0,0 +1,364 @@
+	}
+	return r.Setup(ctx, mgr, bldr)
+}
+
+func (r *ChildSetReconciler[T, CT, CLT]) childReconcilerFor(desired CT, desiredErr error, id string, void bool) *ChildReconciler[T, CT, CLT] {
+	return &ChildReconciler[T, CT, CLT]{
+		Name:               id,
+		ChildType:          r.ChildType,
+		ChildListType:      r.ChildListType,
+		SkipOwnerReference: r.SkipOwnerReference,
+		DesiredChild: func(ctx context.Context, resource T) (CT, error) {
+			return desired, desiredErr
+		},
+		ReflectChildStatusOnParent: func(ctx context.Context, parent T, child CT, err error) {
+			result := retrieveChildSetResult[CT](ctx)
+			result.Children = append(result.Children, ChildSetPartialResult[CT]{
+				Id:    id,
+				Child: child,
+				Err:   err,
+			})
+			stashChildSetResult(ctx, result)
+		},
+		HarmonizeImmutableFields: r.HarmonizeImmutableFields,
+		MergeBeforeUpdate:        r.MergeBeforeUpdate,
+		ListOptions:              r.ListOptions,
+		OurChild: func(resource T, child CT) bool {
+			if r.OurChild != nil && !r.OurChild(resource, child) {
+				return false
+			}
+			return void || id == r.IdentifyChild(child)
+		},
+		Sanitize: r.Sanitize,
+	}
+}
+
+func (r *ChildSetReconciler[T, CT, CLT]) validate(ctx context.Context) error {
+	// default implicit values
+	if r.Finalizer != "" {
+		r.SkipOwnerReference = true
+	}
+
+	// require DesiredChildren
+	if r.DesiredChildren == nil {
+		return fmt.Errorf("ChildSetReconciler %q must implement DesiredChildren", r.Name)
+	}
+
+	// require ReflectChildrenStatusOnParent
+	if r.ReflectChildrenStatusOnParent == nil {
+		return fmt.Errorf("ChildSetReconciler %q must implement ReflectChildrenStatusOnParent", r.Name)
+	}
+
+	if r.OurChild == nil && r.SkipOwnerReference {
+		// OurChild is required when SkipOwnerReference is true
+		return fmt.Errorf("ChildSetReconciler %q must implement OurChild since owner references are not used", r.Name)
+	}
+
+	// require IdentifyChild
+	if r.IdentifyChild == nil {
+		return fmt.Errorf("ChildSetReconciler %q must implement IdentifyChild", r.Name)
+	}
+
+	return nil
+}
+
+func (r *ChildSetReconciler[T, CT, CLT]) Reconcile(ctx context.Context, resource T) (Result, error) {
+	r.init()
+
+	log := logr.FromContextOrDiscard(ctx).
+		WithName(r.Name)
+	ctx = logr.NewContext(ctx, log)
+
+	cr, err := r.composeChildReconcilers(ctx, resource)
+	if err != nil {
+		return Result{}, err
+	}
+	result, err := cr.Reconcile(ctx, resource)
+	r.reflectStatus(ctx, resource)
+	return result, err
+}
+
+func (r *ChildSetReconciler[T, CT, CLT]) composeChildReconcilers(ctx context.Context, resource T) (SubReconciler[T], error) {
+	c := RetrieveConfigOrDie(ctx)
+
+	childIDs := sets.NewString()
+	desiredChildren, desiredChildrenErr := r.DesiredChildren(ctx, resource)
+
+	desiredChildByID := map[string]CT{}
+	for _, child := range desiredChildren {
+		id := r.IdentifyChild(child)
+		if !r.voidReconciler.ourChild(resource, child) {
+			return nil, fmt.Errorf("desired child %q must be claimed by OurChild", id)
```

Comment:
When I test this with my use case, I got this error because I was hoping to rely on the `ownerReference` added by the `ResourceManager` and the default `OurChild`. But this is called before the `ownerReference` gets added.

Reply:
fixed